### ISSUE TITLE
cluster: vmselect query fails sporadically

### ISSUE THREAD

**Issue by salarali** [VictoriaMetrics/VictoriaMetrics#1562]
**Describe the bug**

Sporadically query from the UI through vmselect fails and this is the log I see in vmselect
```
Aug 24 16:51:55 ip-172-30-89-153.ec2.internal vmselect[7433]: 2021-08-24T16:51:55.328Z        warn        VictoriaMetrics/app/vmselect/main.go:523        error in "/select/0/prometheus/api/v1/query?query=sort((time()%20-%20vm_app_start_timestamp%7Baws_name%3D~%22vm-agent-1%22%7D))&time=1629823916": error when executing query="sort((time() - vm_app_start_timestamp{aws_name=~\"vm-agent-1\"}))" for (time=1629823885328, step=300000): cannot evaluate "vm_app_start_timestamp{aws_name=~\"vm-agent-1\"}": error occured during search: cannot fetch query results from vmstorage nodes: cannot perform search on vmstorage 172.30.73.254:8401: cannot execute rpcName="search_v4" on vmstorage "172.30.73.254:8401" with timeout 30.000 seconds (elapsed 0.329 seconds); the timeout can be adjusted with `-search.maxQueryDuration` command-line flag: cannot read error message: cannot read 8 bytes with data size: cannot read data in 0.000 seconds: unexpected EOF; read only 0 bytes
```
Refreshing on grafana returns the query correctly.
At the same time, I have noticed that vmstorage remakes the connections to vmselect
```
Aug 24 16:51:55 ip-172-30-90-1.ec2.internal vmstorage[21193]: 2021-08-24T16:51:55.339Z        info        VictoriaMetrics/app/vmstorage/transport/server.go:164        accepted vmselect conn from 172.30.89.153:47206
Aug 24 16:51:55 ip-172-30-90-1.ec2.internal vmstorage[21193]: 2021-08-24T16:51:55.339Z        info        VictoriaMetrics/app/vmstorage/transport/server.go:164        accepted vmselect conn from 172.30.89.153:47212
Aug 24 16:51:55 ip-172-30-90-1.ec2.internal vmstorage[21193]: 2021-08-24T16:51:55.340Z        info        VictoriaMetrics/app/vmstorage/transport/server.go:164        accepted vmselect conn from 172.30.89.153:47216
Aug 24 16:51:55 ip-172-30-90-1.ec2.internal vmstorage[21193]: 2021-08-24T16:51:55.371Z        info        VictoriaMetrics/app/vmstorage/transport/server.go:164        accepted vmselect conn from 172.30.89.153:47218
Aug 24 16:51:55 ip-172-30-90-1.ec2.internal vmstorage[21193]: 2021-08-24T16:51:55.371Z        info        VictoriaMetrics/app/vmstorage/transport/server.go:207        processing vmselect conn from 172.30.89.153:47216
Aug 24 16:51:55 ip-172-30-90-1.ec2.internal vmstorage[21193]: 2021-08-24T16:51:55.371Z        info        VictoriaMetrics/app/vmstorage/transport/server.go:207        processing vmselect conn from 172.30.89.153:47206
Aug 24 16:51:55 ip-172-30-90-1.ec2.internal vmstorage[21193]: 2021-08-24T16:51:55.371Z        info        VictoriaMetrics/app/vmstorage/transport/server.go:207        processing vmselect conn from 172.30.89.153:47212
Aug 24 16:51:55 ip-172-30-90-1.ec2.internal vmstorage[21193]: 2021-08-24T16:51:55.383Z        info        VictoriaMetrics/app/vmstorage/transport/server.go:164        accepted vmselect conn from 172.30.89.153:47220
Aug 24 16:51:55 ip-172-30-90-1.ec2.internal vmstorage[21193]: 2021-08-24T16:51:55.398Z        info        VictoriaMetrics/app/vmstorage/transport/server.go:207        processing vmselect conn from 172.30.89.153:47218
Aug 24 16:51:55 ip-172-30-90-1.ec2.internal vmstorage[21193]: 2021-08-24T16:51:55.417Z        info        VictoriaMetrics/app/vmstorage/transport/server.go:207        processing vmselect conn from 172.30.89.153:47220
```

**Expected behavior**
No errors when querying the data

**Version**
```
vmselect-20210819-093240-tags-v1.64.1-cluster-0-gee1f3414d
```
```
vmstorage-20210819-093244-tags-v1.64.1-cluster-0-gee1f3414d
```

Let me know if there are any other stats that would help with this query. I am running the basic configs.

**Comment by valyala** [comment]
The error message suggests that the connection between `vmselect` and `vmstorage` unexpectedly breaks. The connection can break due to the following reasons:
* On the crash of vmstorage node. Could you verify whether `vmstorage` nodes crash at the time you see warn messages about broken connections at `vmselect` nodes?
* Due to some timeout. How often do you send queries to `vmselect` nodes? How often do you see the message about broken connections at `vmselect` nodes?
* Due to networking issues between `vmselect` and `vmstorage` nodes.

**Comment by salarali** [comment]
> On the crash of vmstorage node. Could you verify whether vmstorage nodes crash at the time you see warn messages about broken connections at vmselect nodes?

vmstorage nodes dont crash when I see the broken connection error message

> Due to some timeout. How often do you send queries to vmselect nodes? How often do you see the message about broken connections at vmselect nodes?

I am seeing this error when I use grafana, so that usage can be sporadic. Errors when using grafana are also sporadic. Am also making api calls, which makes a call every 5 mins and it sees the error on the first try and then succeeds when I retry it after a second.


I am guessing its because of some timeout, or stale connection issue, because after the first query fails and the connection is remade, the subsequent query succeeds.

**Comment by salarali** [comment]
I also have this sysctl config. Could there be a config that could be effecting the connections?

```
	net.core.netdev_max_backlog = 200000
	net.core.somaxconn = 65535

	net.ipv4.ip_local_port_range = 12284 65535
	net.ipv4.tcp_max_syn_backlog = 65535
	net.ipv4.tcp_slow_start_after_idle = 0
	net.ipv4.tcp_low_latency = 1

	net.core.default_qdisc = fq
	net.ipv4.tcp_congestion_control = bbr
	net.ipv4.tcp_notsent_lowat = 16384

	vm.swappiness = 0

```

**Comment by valyala** [comment]
@salarali , it looks like the issue has been identified and fixed by @mxlxm in the commit 42e07cfaeae842a773933ed0cbabf3cc17c96ce7 . Could you build `vmselect` from this commit according to [these docs](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#building-from-sources) and verify whether this fixes the issue?

**Comment by salarali** [comment]
I have deployed it for a few hours and I have stopped seeing the error. 

Thanks for the fix.
